fdbserver/TagPartitionedLogSystem.actor.cpp

```diff
@@ -562,67 +562,101 @@ struct TagPartitionedLogSystem : ILogSystem, ReferenceCounted<TagPartitionedLogS
 	                              SpanID const& spanContext,
 	                              Optional<UID> debugID,
-	                              ptxn::TLogGroupID tLogGroup) {
+	                              std::vector<ptxn::TLogGroupID> tLogGroups) {
 		vector<Future<Void>> quorumResults;
 		vector<Future<ptxn::TLogCommitReply>> allReplies;
-		int location = 0;
 		Span span("TPLS:push"_loc, spanContext);
+		ASSERT(tLogGroups.size() == prevVersions.size());
+
+		std::unordered_map<UID, vector<ptxn::TLogGroupID>> idToGroupIds;
+		std::unordered_map<UID, vector<Version>> idToPrevVersions;
+
+		// no more than 1 request will be sent to each tlog, even if the tlog belong to multiple tlog groups.
+		// we include all tlog groups data in this aggregated request.
+		// recording the tlog id that already has a request sent, so that no duplicate requests
+		std::unordered_map<UID, int> submitted;
+
 		for (auto& it : tLogs) {
+			// for all logset(each DC in current epoch)
 			if (it->isLocal && it->groupIdToInterfaces.size()) {
-				ASSERT(it->groupIdToInterfaces.count(tLogGroup));
-				const auto& logServers = it->groupIdToInterfaces[tLogGroup];
-				vector<Future<Void>> tLogCommitResults;
-				for (int loc = 0; loc < logServers.size(); loc++) {
-					auto serialized = data.pGroupMessageBuilders->find(tLogGroup)->second->getAllSerialized();
-					allReplies.push_back(logServers[loc]->get().interf().commit.getReply(
-					    ptxn::TLogCommitRequest(spanContext,
-					                            tLogGroup,
-					                            serialized.first,
-					                            serialized.second,
-					                            prevVersion,
-					                            version,
-					                            knownCommittedVersion,
-					                            minKnownCommittedVersion,
-					                            debugID),
-					    TaskPriority::ProxyTLogCommitReply));
-					Future<Void> commitSuccess = success(allReplies.back());
-					addActor.get().send(commitSuccess);
-					tLogCommitResults.push_back(commitSuccess);
-					location++;
+				for (int i = 0; i < tLogGroups.size(); i++) {
+					ASSERT(it->groupIdToInterfaces.count(tLogGroups[i]));
+					const auto& logServers = it->groupIdToInterfaces[tLogGroups[i]];
+					for (int loc = 0; loc < logServers.size(); loc++) {
+						UID id = logServers[loc]->get().id();
+						idToGroupIds[id].push_back(tLogGroups[i]);
+						idToPrevVersions[id].push_back(prevVersions[i]);
+					}
+				}
+				for (int i = 0; i < tLogGroups.size(); i++) {
+					ASSERT(it->groupIdToInterfaces.count(tLogGroups[i]));
+					const auto& logServers = it->groupIdToInterfaces[tLogGroups[i]];
+					vector<Future<ptxn::TLogCommitReply>> tLogCommitResultsForGroup;
+					for (int loc = 0; loc < logServers.size(); loc++) {
+						UID id = logServers[loc]->get().id();
+						if (submitted.find(id) != submitted.end()) {
+							tLogCommitResultsForGroup.push_back(allReplies[submitted[id]]);
```

Comment:
If `id` exists, then I want to use the same future, is it the right way to do so?